README.md
# Fermat

[![CircleCI](https://dl.circleci.com/status-badge/img/gh/JordanRL/Fermat/tree/master.svg?style=shield)](https://dl.circleci.com/status-badge/redirect/gh/JordanRL/Fermat/tree/master) [![Test Coverage](https://api.codeclimate.com/v1/badges/ce0c5761a8f0d3d37cf2/test_coverage)](https://codeclimate.com/github/JordanRL/Fermat/test_coverage) [![Maintainability](https://api.codeclimate.com/v1/badges/ce0c5761a8f0d3d37cf2/maintainability)](https://codeclimate.com/github/JordanRL/Fermat/maintainability) [![Latest Stable Version](https://poser.pugx.org/samsara/fermat/v/stable)](https://packagist.org/packages/samsara/fermat) [![Total Downloads](https://poser.pugx.org/samsara/fermat/downloads)](https://packagist.org/packages/samsara/fermat) [![License](https://poser.pugx.org/samsara/fermat/license)](https://packagist.org/packages/samsara/fermat)

**This project is unit tested against 8.1, and merges are not accepted unless the tests pass.**

## Installation

To install, simply require the package using composer:

    composer require "samsara/fermat:^2.1"
    
Or include it in your `composer.json` file:

```json
{
    "require": {
        "samsara/fermat": "^2.1"
    }
}
```

The project namespace is `Samsara\Fermat\Core\*`. You can view the project on [Packagist](https://packagist.org/packages/samsara/fermat).

### Modules

**NOTE:** For the upcoming 3.0 release, modules will once again be bundled.

Modules for Fermat provide additional functionality, as many of these features would be unused by most people.

- [Algebra Expressions](https://github.com/SamsaraLabs/FermatAlgebraExpressions): Provides support for algebraic expressions such as polynomials and functions.
- [Complex Numbers](https://github.com/SamsaraLabs/FermatComplexNumbers): Provides complex numbers and enables additional features in Decimal instances.
- [Coordinate Systems](https://github.com/SamsaraLabs/FermatCoordinateSystems): Provides coordinate systems for cartesian, spherical, polar, and cylindrical coordinates.
- [Matrices & Vectors](https://github.com/SamsaraLabs/FermatMatricesAndVectors): Provides support for matrix math and vector math.
- [Statistics](https://github.com/SamsaraLabs/FermatStats): Provides support for statistical operations and distributions.

To require the entire library, including all available modules, use:

    composer require samsara/fermat-all "^1.0"

Or in your `composer.json` file:

```json
{
    "require": {
        "samsara/fermat-all": "^1.0"
    }
}
```

## Documentation

The `Samsara\Fermat\Core\Numbers` factory class provides a way to use the Value classes in Fermat without being as specific as those classes may require. Consider the following code:

```php
<?php

use Samsara\Fermat\Core\Core\Numbers;

$five = Numbers::make(Numbers::IMMUTABLE, 5);
$ten = Numbers::make(Numbers::IMMUTABLE, '10');

echo $five->add($ten); // Prints: "15"
```

Note that the `make()` method allows you to provide both an int and a string as the value. In fact, it also allows you to provide a float. The first argument is the specific class that will be used for the value, the second argument is the value itself. The third and fourth arguments are optional and represent the scale (in number of decimal places) and the base of the number respectively. The scale and base arguments will only accept integer values.

If you do not specify a scale value, and you are using the default values, it automatically has a scale of either 10, or the string length of the input value, whichever is greater.

Here is an example of using the factory method to make a value that is in a base other than base10:

```php
<?php

use Samsara\Fermat\Core\Core\Numbers;
use Samsara\Fermat\Core\Core\Enums\NumberBase;

// Value in base5
$five = Numbers::make(Numbers::IMMUTABLE, '10', null, NumberBase::Five); 
// Value in base10
$ten = Numbers::make(Numbers::IMMUTABLE, '10'); 

echo $ten->add($five); // Prints: "15" (The sum in base10)
echo $five->add($ten); // Prints: "30" (The sum in base5)
```

You can also use a `Fraction` and `Number` together:

```php
<?php

use Samsara\Fermat\Core\Core\Values\ImmutableDecimal;
use Samsara\Fermat\Core\Core\Values\ImmutableFraction;

$five = new ImmutableDecimal(5);
$oneQuarter = new ImmutableFraction(1, 4);

echo $five->add($oneQuarter); // Prints: "5.25"
// The asDecimal() method is called on $oneQuarter

echo $oneQuarter->add($five); // Prints: "21/4"
// Calls getValue() on $five and instantiates a new ImmutableFraction
```

You can read the full documentation for Fermat [here](https://jordanrl.github.io/Fermat/).

## Contributing

Please ensure that pull requests meet the following guidelines:

- New files created in the pull request must have a corresponding unit test file, or must be covered within an existing test file.
- Your merge may not drop the project's test coverage below 80%.
- Your merge may not drop the project's test coverage by MORE than 5%.
- Your merge must pass CI build tests for PHP 8.1.

For more information, please see the section on [Contributing](CONTRIBUTING.md)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/). At least so much as we feel like it.

## [Unreleased] 
## Included In This Release

### Features

- Added `RoundingProvider` and New Rounding Modes @JordanRL (#115)
- Added `subFactorial()` method @JordanRL (#114)

### Maintenance

- Added `RoundingProvider` and New Rounding Modes @JordanRL (#115)
- Added `subFactorial()` method @JordanRL (#114)


## [2.1.1] - 2021-06-29
### Added
- **RandomProvider**: A provider to supply random numbers.
  - *randomInt()*: Supplies a random integer.
  - *randomDecimal()*: Supplies a random decimal between 0 and 1 (inclusive)
  - *randomReal()*: Supplies a random real number.

### Changed
- **Composer**: Removed dependency on ircmaxell/random-lib, as it was outdated and is now replaced with Fermat's RandomProvider.

## [2.1.0] - 2021-06-27
### Added
- NumberCollection: Added the method `selectScale()` to return the int value of the largest scale among all collection members.

## [2.0.1] - 2021-06-26
### Fixed
- Composer: Fixed an issue with dependencies that created errors for libraries which include this as a dependency.

## [2.0.0] - 2021-06-26
### Added

- Providers:
  - Distributions:
    - Base: Distribution abstract class
  - ArithmeticProvider:
    - Added a scale argument to all methods that were missing it.
  - PolyfillProvider
  - SequenceProvider:
    - `nthFibonacciNumber` function
  - StatsProvider:
    - Added docblocks for all methods
    - Added `gammaFunction` function
- Types:
  - Base:
    - Interfaces:
        - BaseConversionInterface
        - ComplexInterface
        - DecimalInterface:
          - Added hyperbolic trig functions
          - Added `exp()`
        - NumberInterface:
          - Added `isComplex(): bool` method
        - MatrixInterface
        - NumberCollectionInterface
    - Number:
      - numberOfTotalDigits()
      - numberOfIntDigits()
      - numberOfDecimalDigits()
      - numberOfSigDecimalDigits()
      - isComplex()
      - asComplex()
  - Traits:
    - ArithmeticTrait
    - ComparisonTrait
    - IntegerMathTrait
    - InverseTrigonometryTrait
    - LogTrait
    - ScaleTrait
    - TrigonometryTrait:
      - Added implementation of hyperbolic trig functions
  - ComplexNumber
  - Coordinate
  - Decimal
  - Expression
  - Matrix
  - NumberCollection
- Values:
  - Algebra
    - PolynomialFunction
  - Geometry
    - CoordinateSystems
      - CartesianCoordinate
      - CylindricalCoordinate
      - PolarCoordinate
      - SphericalCoordinate
  - ImmutableComplexNumber
  - ImmutableFraction:
    - Added implementation of `isComplex(): bool`
  - ImmutableMatrix
  - ImmutableNumber:
    - Added implementation of `isComplex(): bool`
  - MutableComplexNumber
  - MutableFraction:
    - Added implementation of `isComplex(): bool`
  - MutableMatrix
  - MutableNumber:
    - Added implementation of `isComplex(): bool`
- Factories:
  - Collections
  - ComplexNumbers
  - Matrices

### Removed

- Composer: 
  - **Suggested:**
    - ext-stats: * (As this library is no longer maintained and unavailable on PHP 7+, it is being removed)
- Providers:
  - SequenceProvider
    - Removed empty `nthSecTanCoefNumber` function until an implementation exists

### Changed

- Providers:
  - Distributions:
    - Removed references to ext-stats in all distributions
    - Fixed missing @throws in docblocks in all distributions
    - Now extend `Distribution` abstract class
    - Use the randomInt() method in PolyfillProvider instead of RandomLib directly
  - SequenceProvider:
    - Allowed `nthEvenNumber()`, `nthOddNumber()`, `nthPowerNegativeOne()`, `nthEulerZigzag()`, and `nthFibonacciNumber()` to return windows of the sequence as a NumberCollection object
- Types:
  - Base:
    - Interfaces:
      - ALL: Changed namespace from `Samsara\Fermat\Core\Types\Base` to `Samsara\Fermat\Core\Types\Base\Interfaces`
      - DecimalInterface:
        - Changed signature of `log10()` and `ln()` to `$scale = null` instead of `$scale = 10`
        - Changed type of `$scale` parameter in `log10()` and `ln()` from `int` to `int|null`
      - NumberInterface:
        - Moved the `convertToBase()` method to the new `BaseConversionInterface`
    - Traits:
      - ArithmeticTrait:
        - Updated all arithmetic functions to work with imaginary numbers, complex numbers, and negative numbers (square root)
    - Number:
      - Changed namespace from `Samsara\Fermat\Core\Types` to `Samsara\Fermat\Core\Types\Base`
      - Moved many methods into traits
      - Now passes the scale setting to ALL calls to `ArithmeticProvider` allowing scale up to 2147483646 digits.
      - Trig functions are now arbitrary scale
      - numberOfLeadingZeros() now returns `int` type, as was originally intended
  - Fraction:
    - Moved many methods into traits
  

### Fixed

- Providers:
  - Distributions:
    - Exponential:
      - Fixed accidental cast to float in `random` function
  - SequenceProvider:
    - Fixed erroneous limit on Euler Zigzag sequence index (45 -> 50) and updated exception to reflect current limit (39 -> 50)
- Types:
  - Traits:
    - LogTrait:
      - Fixed `$scale` to work the way it does in other functions for `log10()` and `ln()`
  - Tuple:
    - Changed the constructor so that it properly works with either parameter collection or a single array
  - ImmutableNumber:
    - Fixed the problem with scale in continuousModulo()
  - MutableNumber:
    - Fixed the problem with scale in continuousModulo()

## [1.0.2] - 2017-01-16
### Fixed
- DecimalInterface
  - Added the following to correctly reflect usage:
    - tan()
    - cot()
    - sec()
    - csc()
    - arcsin()
    - arccos()
    - arctan()
    - arccot()
    - arcsec()

src/Samsara/Cantor/Bases.php
<?php

namespace Samsara\Cantor;

use Samsara\Cantor\Provider\BCProvider;

class Bases
{

    public static $chars = [
        '0' => 0,
        '1' => 1,
        '2' => 2,
        '3' => 3,
        '4' => 4,
        '5' => 5,
        '6' => 6,
        '7' => 7,
        '8' => 8,
        '9' => 9,
        'A' => 10,
        'B' => 11,
        'C' => 12,
        'D' => 13,
        'E' => 14,
        'F' => 15,
        'G' => 16,
        'H' => 17,
        'I' => 18,
        'J' => 19,
        'K' => 20,
        'L' => 21,
        'M' => 22,
        'N' => 23,
        'O' => 24,
        'P' => 25,
        'Q' => 26,
        'R' => 27,
        'S' => 28,
        'T' => 29,
        'U' => 30,
        'V' => 31,
        'W' => 32,
        'X' => 33,
        'Y' => 34,
        'Z' => 35
    ];

    public static function charFromVal($val)
    {
        if ($val > 35 || $val < 0) {
            throw new \InvalidArgumentException('Only bases up to base 36 are supported.');
        }

        return array_search((int)$val, self::$chars);
    }

    public static function valFromChar($char)
    {
        $char = strtoupper($char);

        if (array_key_exists($char, self::$chars)) {
            return self::$chars[$char];
        } else {
            throw new \InvalidArgumentException('Only alphanumeric characters are valid numbers.');
        }
    }

    public static function doBaseConvMath($num, $oldBase, $newBase)
    {
        $numArr = array_reverse(str_split($num));

        $i = 1;
        $d = 0;

        $sum = '0';

        foreach ($numArr as $digit) {
            $sum = BCProvider::add($sum, BCProvider::multiply(
                Bases::valFromChar($digit),
                BCProvider::exp($oldBase, $d)
            ));
            $d++;
        }

        $converted = '';
        do {
            if (BCProvider::compare(0, $sum) == 0 || BCProvider::compare(0, $sum) == 1) {
                break;
            }
            $newVal = BCProvider::modulo($sum, BCProvider::exp($newBase, $i));
            $newDigit = Bases::charFromVal(BCProvider::divide($newVal, BCProvider::exp($newBase, $i-1)));
            $sum = BCProvider::subtract($sum, $newVal);
            $converted = $newDigit . $converted;
            $i++;
        } while (true);

        return $converted;
    }

    public static function doFractBaseConvMath($num, $oldBase, $newBase)
    {
        // TODO: Do
    }

}